File: archrules-common/src/main/java/com/netflix/nebula/archrules/common/KotlinInternalClassPredicate.java
package com.netflix.nebula.archrules.common;

import com.tngtech.archunit.base.DescribedPredicate;
import com.tngtech.archunit.core.domain.JavaClass;
import kotlin.Metadata;
import kotlin.metadata.Attributes;
import kotlin.metadata.Visibility;
import kotlin.metadata.jvm.KotlinClassMetadata;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

@NullMarked
public class KotlinInternalClassPredicate extends DescribedPredicate<JavaClass> {
    public KotlinInternalClassPredicate() {
        super("Kotlin internal class");
    }

    @Override
    public boolean test(@Nullable JavaClass javaClass) {
        if (javaClass != null && javaClass.isAnnotatedWith("kotlin.Metadata")) {
            KotlinClassMetadata metadata = KotlinClassMetadata.readStrict(javaClass.getAnnotationOfType(Metadata.class));
            if (metadata instanceof KotlinClassMetadata.Class) {
                return Attributes.getVisibility(((KotlinClassMetadata.Class) metadata).getKmClass()) == Visibility.INTERNAL;
            } else {
                return false;
            }
        } else {
            return false;
        }
    }
}

File: archrules-common/src/main/kotlin/com/netflix/nebula/archrules/common/KotlinInternalClassPredicate.kt
package com.netflix.nebula.archrules.common

import com.tngtech.archunit.base.DescribedPredicate

class KotlinInternalClassPredicate : DescribedPredicate<com.tngtech.archunit.core.domain.JavaClass>(
    "Kotlin internal class"
) {
    /**
     * @param javaClass declared as nullable in order to avoid static references to the kotlin Intrinsics class
     */
    override fun test(javaClass: com.tngtech.archunit.core.domain.JavaClass?): Boolean {
        return javaClass != null && javaClass.isAnnotatedWith("kotlin.Metadata") &&
            javaClass.reflect().kotlin.visibility.toString() == "INTERNAL"
    }
}
